spring-integration-core/src/main/java/org/springframework/integration/support/converter/SimpleMessageConverter.java

```diff
@@ -103,7 +105,14 @@ protected MessageBuilderFactory getMessageBuilderFactory() {
 	@Override
 	public Message<?> toMessage(Object object, MessageHeaders headers) {
 		try {
-			return this.inboundMessageMapper.toMessage(object);
+			Message message = this.inboundMessageMapper.toMessage(object);
+			if (headers != null) {
+				message =
+						MessageBuilder.fromMessage(message)
+								.copyHeadersIfAbsent(headers)
+								.build();
+			}
```

Comment:
Hmmm - this will impact every messaging gateway. I wonder if we could have the mapper return the builder (perhaps a `default` method)?

Or provide an overloaded method that takes extra headers.

Or create a new converter that's only used by Redis.

Reply:
As you see I have switched `RedisInboundChannelAdapter` to use `org.springframework.messaging.converter.SimpleMessageConverter` instead.

But I agree, I'm looking now to something "not very impacting"...